rfcs/20200505-transactional-fs.md

```diff
@@ -0,0 +1,700 @@
+    status = AppendToFile(output, inp);
+    if (!status.ok()) return status;
+  }
+  return status;
+}
+```
+
+This example would be modified as below to work with transactions as described in this proposal.
+
+```cpp
+Status MergeFiles(const string& fname, const string& dirname,
+                  const vector<string> input_files) {
+  TransactionToken* token = nullptr;
+  auto status = Env::Default()->StartTransaction(fname, &token);
+  if (!status.ok()) {
+    LOG(WARNING) << "Starting transaction for " << fname << " failed with \""
+                 << status << "\". Continuing without transactions";
+  }
+  status = Env::Default()->IsDirectory(dirname, token);
+  if (!status.ok()) {
+    status = Env::Default()->CreateDirectory(dirname, token);
+    if (!status.ok()) {
+      return status;
+    }
+  }
+  std::unique_ptr<WritableFile> output;
+  status = Env::Default()->NewAppendableFile(fname, &output, token);
+  if (!status.ok()) return status;
+  for (const auto& inp : input_files) {
+    // read file inp and append to output after processing it
+    status = AppendToFile(output, inp);
+    if (!status.ok()) return status;
+  }
+  return status;
+}
+```
+
+With the Wrapped filesytems proposal it would be like
+
+```cpp
+Status MergeFiles(const string& fname, const string& dirname,
+                  const vector<string> input_files) {
+  std::unique_ptr<Filesystem> FS;
+  auto status = Env::Default()->StartTransactionForURI(fname, &FS);
+  if (!status.ok()) {
+    LOG(WARNING) << "Starting transaction for " << fname << " failed with \""
+                 << status << "\". Continuing without transactions";
+  }
+  status = FS->IsDirectory(dirname);
+  if (!status.ok()) {
+    status = FS->CreateDirectory(dirname);
+    if (!status.ok()) {
+      return status;
+    }
+  }
+  std::unique_ptr<WritableFile> output;
+  status = FS->NewAppendableFile(fname, &output);
+  if (!status.ok()) return status;
+  for (const auto& inp : input_files) {
+    status = AppendToFile(output, inp);
+    if (!status.ok()) return status;
+  }
+  return status;
+}
+```
+
+And with the stateful tokens proposal would be
+
+```cpp
+Status MergeFiles(const string& fname, const string& dirname,
+                  const vector<string> input_files) {
+  TransactionToken* token = nullptr;
+  auto status = Env::Default()->GetTokenForURI(fname, &token);
+  if (!status.ok()) {
+    LOG(WARNING) << "Starting transaction for " << fname << " failed with \""
+                 << status << "\". Continuing without transactions";
+  }
+  Env::Default()->AddToTransaction(token, dirname);
+  status = Env::Default()->IsDirectory(dirname, token);
+  if (!status.ok()) {
+    status = Env::Default()->CreateDirectory(dirname, token);
+    if (!status.ok()) {
+      return status;
+    }
+  }
+  Env::Default()->AddToTransaction(token, fname);
+  std::unique_ptr<WritableFile> output;
+  status = Env::Default()->NewAppendableFile(fname, &output, token);
```

Comment:
```suggestion
  status = Env::Default()->NewAppendableFile(fname, &output);
```

Reply:
Sorry for that.  I forgot to remove them during copy-paste.